subprojects/version-control/src/main/java/org/gradle/vcs/internal/DefaultVersionControlSystemFactory.java

```diff
@@ -16,17 +16,66 @@
 
 package org.gradle.vcs.internal;
 
+import org.gradle.cache.CacheRepository;
+import org.gradle.cache.FileLockManager;
+import org.gradle.cache.PersistentCache;
 import org.gradle.vcs.VersionControlSpec;
 import org.gradle.vcs.VersionControlSystem;
+import org.gradle.vcs.VersionRef;
+import org.gradle.vcs.git.GitVersionControlSpec;
 import org.gradle.vcs.git.internal.GitVersionControlSystem;
 
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
+
 public class DefaultVersionControlSystemFactory implements VersionControlSystemFactory {
+    private final CacheRepository cacheRepository;
+    private final Map<Class<? extends VersionControlSpec>, VersionControlSystem> versionControlSystems;
+
+    DefaultVersionControlSystemFactory(CacheRepository cacheRepository) {
+        this.cacheRepository = cacheRepository;
+        // TODO: Move this map to a registry and inject the registry.
+        this.versionControlSystems = new HashMap<Class<? extends VersionControlSpec>, VersionControlSystem>();
+        versionControlSystems.put(DirectoryRepository.class, new SimpleVersionControlSystem());
+        versionControlSystems.put(GitVersionControlSpec.class, new GitVersionControlSystem());
+    }
+
     @Override
     public VersionControlSystem create(VersionControlSpec spec) {
-        // TODO: Register these mappings somewhere
-        if (spec instanceof DirectoryRepository) {
-            return new SimpleVersionControlSystem();
+        return new ThreadSafeVersionControlSystem(versionControlSystems.get(spec.getClass()), cacheRepository);
```

Comment:
So after seeing this, I wonder if we should tweak the `VersionControlSystem` interface to require that implementors have a constructor that takes a `VersionControlSpec` and then remove `VersionControlSpec` from all of the methods in `VersionControlSystem`. We would then create a new `VersionControlSystem` every time we needed to access a given repo (defined by a `VersionControlSpec`).  WDYT?

Reply:
I don't understand what problem this would solve for us. Doesn't it just require more memory allocation and garbage collection?

Reply:
Potentially that could be a downside.  I was just thinking about the API turning into this:

```
interface VersionControlSystem {
    Set<VersionRef> getAvailableVersions();
    void populate(File workingDir, VersionRef ref);
}
```

There would be ways to cache the `VersionControlSystem`s per `VersionControlSpec`, but it's not obviously better than what we have now.

Reply:
I am more comfortable leaving the interface the way it is right now. We can discuss the pros and cons separately, but I don't see any compelling reason to maintain more objects at this point.